app/src/main/java/org/openobservatory/ooniprobe/data/TestData.java

```diff
@@ -313,6 +347,113 @@ public static void setAnomaly_ndt(String entry, NetworkMeasurement test){
+        }
+        try {
+            int anomaly = OONITests.ANOMALY_GREEN;
+            JSONObject jsonObj = new JSONObject(entry);
+            String keys[] = {"whatsapp_endpoints_status",
+                    "whatsapp_web_status",
+                    "registration_server_status"};
+            for (String key: keys)
+            {
+                String value = jsonObj.getString(key);
+                if (value == null)
+                    anomaly = OONITests.ANOMALY_ORANGE;
+                else if (value.equals("blocked"))
+                    anomaly = OONITests.ANOMALY_RED;
+            }
+            if (test.anomaly < anomaly) {
+                test.anomaly = anomaly;
+                TestStorage.setAnomaly(context, test.test_id, anomaly);
+            }
+        } catch (JSONException e) {
+            System.out.println("JSONException "+ e);
+        }
+    }
+
+    /*
+    for telegram: red if either "telegram_http_blocking" or "telegram_tcp_blocking" is true, OR if ""telegram_web_status" is "blocked"
+    the "*_failure" keys for telegram and whatsapp might indicate a test failure / anomaly
+    docs: https://github.com/TheTorProject/ooni-spec/blob/master/test-specs/ts-020-telegram.md#semantics
+    */
+    public static void setAnomaly_telegram(String entry, NetworkMeasurement test) {
+        if(!test.entry) {
+            TestStorage.setEntry(context, test);
+            test.entry = true;
+        }
+        try {
+            int anomaly = OONITests.ANOMALY_GREEN;
+            JSONObject jsonObj = new JSONObject(entry);
+            String keys[] = {"telegram_http_blocking",
+                    "telegram_tcp_blocking"};
+            for (String key: keys)
+            {
+                Object value = jsonObj.get(key);
+                Boolean boolvalue = jsonObj.getBoolean(key);
+                if (value == null)
+                    anomaly = OONITests.ANOMALY_ORANGE;
+                else if (boolvalue)
+                    anomaly = OONITests.ANOMALY_RED;
+            }
+            if (jsonObj.has("telegram_web_status")) {
+                String telegram_web_status = jsonObj.getString("telegram_web_status");
+                if (telegram_web_status == null)
+                    anomaly = OONITests.ANOMALY_ORANGE;
+                else if (telegram_web_status.equals("blocked"))
+                    anomaly = OONITests.ANOMALY_RED;
+            }
+            if (test.anomaly < anomaly) {
+                test.anomaly = anomaly;
+                TestStorage.setAnomaly(context, test.test_id, anomaly);
+            }
+        } catch (JSONException e) {
+        }
+
+
+    }
+
+    /*
+    FB: red blocking if either "facebook_tcp_blocking" or "facebook_dns_blocking" is true
+    docs: https://github.com/TheTorProject/ooni-spec/blob/master/test-specs/ts-019-facebook-messenger.md#semantics
+    */
+    public static void setAnomaly_facebook_messenger(String entry, NetworkMeasurement test) {
```

Comment:
The three new functions are all similar. I suggest considering writing a single function with more parameters, as it seems the main difference is the list of keys to check for.

Reply:
These have been tested and working, I suggest to don't touch them for this release as in the next everything is gonna be changed

Reply:
I must admin that I am not comfortable with:

1. blocking someone else work because of suggested refactorings (hence, I'd say let's go forward with this code and okay for improve stuff in next release)

2. the concept that _everything_ is changed by next release (I would be happier to see more incremental improvement where possible; that said, I also understand that apps are very tightly coupled to the UI... but still rewriting the world makes me nervous)